### DISCUSSION TITLE
How we want to work together

### DISCUSSION THREAD

**Discussion by squidfunk** [mkdocs/mkdocs#3672] [General]
We want to see the ecosystem flourish and to push this project to new heights. However, MkDocs had some problems with toxic behavior in the past, which is why I propose to adopt the following manifesto:

---

**Manifesto for MkDocs Plugin and Extension Authors Community**

Welcome to our community! We believe in fostering a collaborative, friendly, and non-toxic environment where MkDocs plugin and extension authors can thrive. Here's our manifesto:

1. **Inclusivity**: We embrace diversity in all its forms. Everyone is welcome regardless of race, gender, sexual orientation, religion, or background.

2. **Respect**: We treat each other with kindness, empathy, and respect. We value differing opinions and engage in constructive discussions.

3. **Collaboration**: We believe in the power of collaboration. Together, we can achieve more. Let's share knowledge, help each other, and collaborate on projects to improve MkDocs for everyone.

4. **Openness**: We encourage open communication and transparency. Share your ideas, feedback, and concerns openly and respectfully. We're here to listen and support each other.

5. **Non-Toxic Environment**: We have zero tolerance for toxic behavior, including harassment, bullying, discrimination, or any form of abuse. Such behavior will not be tolerated and may result in immediate removal from the community.

6. **Learning and Growth**: We're committed to continuous learning and growth. Let's support each other's development by sharing resources, providing feedback, and celebrating achievements.

7. **Ownership and Accountability**: We take ownership of our actions and their consequences. If we make mistakes, we own up to them, learn from them, and strive to do better.

8. **Fun and Enjoyment**: Lastly, let's have fun! Building plugins and extensions for MkDocs should be an enjoyable experience. Let's celebrate our successes, share our passions, and enjoy the journey together.

By adhering to these principles, we create a welcoming and supportive community where all members can thrive and contribute positively to the MkDocs ecosystem.

---

The most important part is 5. – zero tolerance for toxic behavior. Collaboration and cooperation should be exclusively constructive and non-personal. There's absolutely no reason to move things onto a personal level. In case of violations, I propose to be very strict and adopt [the approach we use over at Material for MkDocs](https://squidfunk.github.io/mkdocs-material/contributing/#fn:1) very successfully:

> Warning and blocking policy: Given the increasing popularity of our project and our commitment to a healthy community, we've defined clear guidelines on how we proceed with violations:
>
> 1. First warning: Users displaying repeated inappropriate, offensive, or harmful behavior will receive a first warning. This warning serves as a formal notice that their behavior is not in alignment with our community standards and Code of Conduct. The first warning is permanent.
>
> 2. Second warning and opportunity for resolution: If the behavior persists, a second warning will be issued. Upon receiving the second warning, the user will be given a 5-day period for reflection, during which they are encouraged to publicly explain or apologize for their actions. This period is designed to offer an opportunity for openly clearing out any misunderstanding.
>
> 3. Blocking: Should there be no response or improvement in behavior following the second warning, we reserve the right to block the user from the community and repository. Blocking is considered a last resort, used only when absolutely necessary to protect the community's integrity and positive atmosphere.
> 
> Blocking has been an exceptionally rare necessity in our overwhelmingly positive community, highlighting our preference for constructive dialogue and mutual respect. It aims to protect our community members and team. [↩](https://squidfunk.github.io/mkdocs-material/contributing/#fnref:1)

We had to use these guidlines only a handful of times, but in the majority of cases, after the first warning, the behavior stopped.

**Comment by pawamoy** [comment]
The manifesto looks great, thanks. If we are to follow a similar approach as the one you mentioned, it will have to be included in the manifesto's contents of course. If the experience with Material for MkDocs shows that it works, I'm OK to try it here too :slightly_smiling_face:

**Reply by squidfunk** [reply]
Yup, we should somehow marry the manifesto with a code of conduct that we need to establish. However, I first wanted to run things by everyone before starting to work on this.

**Reply by waylan** [reply]
There is mention of a code of conduct in the [Contributing Guide](https://github.com/mkdocs/mkdocs/blob/master/CONTRIBUTING.md#code-of-conduct), but that just links out to the [PyPA Code of Conduct](https://www.pypa.io/en/latest/code-of-conduct/). IIRC, back when that was added, the PyPA  Code of Conduct was based on an older version (maybe 1.2 or 1.3) of the [Contributor Covenant](https://www.contributor-covenant.org/). However, it now just links to the [PSF Code of Conduct](https://www.python.org/psf/conduct/), which is very different in certain respects and doesn't seem appropriate for a project like MkDocs (it contains whole sections on in-person events among other things). So, yeah, MkDocs should have its own Code of Conduct.  And this proposal is a great start.

I am especially glad to see item 5. The trick will be to balance that with item 2. We have all seen occasions where individuals have conflated respectful disagreement with bullying.  We need to maintain an environment where everyone can openly state their opinion without fear of retribution.

**Reply by squidfunk** [reply]
Thanks, @waylan, I fully agree. This is just a first draft and we should weave this into a full-blown code of conduct. As said, I first wanted to get some feedback until we decide on what goes into it, and non-toxicity is the absolute most important thing, IMHO.

**Reply by waylan** [reply]
And we definitely need a defined process for enforcement. Back when I was lead, there were occasions where I saw violations, but without a formal process for handing them, I didn't do anything... and that eventually developed into the toxin environment which pushed me away. I was curious what prompted this discussion and see the new changes (congrats to the new team members - they are excellent additions ❤️ ). If that toxic environment can be eliminated, I would be willing to turn notifications back on for the main project. My time is sparse these days, but I'm not opposed to help with triage.

**Reply by byrnereese** [reply]
Not to be too picky... but if mkdocs/mkdocs#3666 is the most important value, then shouldn't it be in the mkdocs/mkdocs-team-space#1 spot? :)

That being said, I think it is important to have articulated behaviors that are not acceptable, with clear consequences for violating that rule. But, if our community is open, inclusive, collaborative, and respectful... then do we need to say "don't be toxic?" I only suggest that because in my experience and opinion, values such as the ones above should always be expressed in the positive. Saying we "don't want x, y or z" frames values in the negative, which casts a subtle shadow on the values as a whole.

**Reply by Andre601** [reply]
Honestly, regarding byrnereese's point, I feel like that 1, 2 and 5 cover what is effectively a "Don't be toxic" rule here...
I mean, if you're harassing someone, act agressive, racist, etc, you effectively act toxic here... Tho I assume they are separated for a higher level of clarity here and to avoid people trying to find loop holes à la "Rules don't state X".

**Reply by pawamoy** [reply]
Interesting, thanks. I think the emphasis is on the zero-tolerance part. It's hard to include "zero-tolerance for X" in a point that focuses on positive things (kindness, respect, etc.). I feel like it is easier to convey it with a separate point. But I agree that it should be clearly defined (particularly "any form of abuse").

**Reply by byrnereese** [reply]
I would add, that we should be clear that any determination of one's adherence to this policy is at the sole discretion of the maintainers of the project. People have to understand that while we strive to conduct ourselves in a fair, just, open and transparent way, on this subject we are not a democracy. 

And we need to preserve our ability to act decisively to protect members of the community, our reputation and our values.

**Reply by squidfunk** [reply]
I'd be honest here: for the most part, this was generated by ChatGPT 😅 I just wanted something that we can discuss, and I agree that we should consolidate those "8 commandments" into 3-4 more crisp rules that we can enforce.

**Reply by Andre601** [reply]
I would propose something alongside this:

1. **Overall respect for one another**
  Different people have different views and histories. As such should you have resoect for ones gender, sexual orientation, race, religious views and similar. Any toxic behaviour such as targeted harassment or attacks on the mentioned points will not be tolerated.
2. **Collaboarion**
  We want to foster a helpful, collaborative community. As such, if you decide to help someone, do so with respect and patience. Do not blame them for lack of knowledge or understanding.
3. **Owning up to your mistakes**
  Making mistakes is normal and can help in improving one self. If someone made a mistake, do not blame them or otherwise make them feel guilty or bad. Instead, help them improve. If you yourself made a mistake, own up to it, so that it can be fixed.
4. **Have fun**
  Working together should be fun. Don't force yourself or others into doing things you or they would not enjoy.

May need some tweaks, but I feel like it would overall condense these 8 points into 4 sets that should cover the most important parts of keeping a collaborative environment.

**Reply by squidfunk** [reply]
IMHO, sexual orientation, gender, race, religion should not be part of any discussion – it should not be personal at any time. MkDocs historically had problems with diverging opinions. From my viewpoint, this is the most important point we should address when talking about respect, shortly mentioning that all personal things should not be a topicalized at all.

Additionally, to keep things manageable for us, we should include that if you make a suggestion in this team space, you actively state that you would like to implement it or at least invest significant time to help out on it. We should not let this discussion board turn into a wishlist of things that "would be awesome". We should keep the focus sharp, otherwise we'd be discussing castles in the sky for eternity, which eats up our time to move forward with the stuff that absolutely needs to be done ☺️

**Reply by Andre601** [reply]
> IMHO, sexual orientation, gender, race, religion should not be part of any discussion

The problem is not whether they are part of a discussion, but rather if someone targets someone because of those.

I've seen it before on communities where people where harassed and attacked just because someone disagreed with their sexual orientation, despite this never having been mentioned nor been made the topic of that place.
So this is more just to have a solid background for why targeted harassment on someones gender, etc is not tolerated. Else we get the usual "The rules don't state X, so I didn't know" excuses... And trust me, people like to use that excuse a lot.

> Additionally, to keep things manageable for us, we should include that if you make a suggestion in this team space, you actively state that you would like to implement it or at least invest significant time to help out on it. We should not let this discussion board turn into a wishlist of things that "would be awesome". We should keep the focus sharp, otherwise we'd be discussing castles in the sky for eternity, which eats up our time to move forward with the stuff that absolutely needs to be done ☺️

I'm mixed on that. On one hand can I understand the reasoning behind this, but on the other can I see situations where someone has an idea, that could also be added, but lacks the knowledge to really add it themself, while others may have the ability too.
Or they just want to spark a discussion on a topic that may be important for the future, but not yet necessary for the project itself (therefore not require active implementation of any kind)...
We shouldn't really stop the posability of brainstorming ideas here, if the topic may not yet be something for the public.

**Reply by squidfunk** [reply]
This is a private space, which includes a very small subset of users of MkDocs, so I think we should concentrate on the actual problems we had in the past maintaining this project. Thus, we can use [GitHub's common Code of Conduct template](https://github.com/squidfunk/mkdocs-material/blob/master/CODE_OF_CONDUCT.md) and refer to that, so that everybody has to abide by that as well. IMHO, our manifesto should focus on how we want to work.

> I'm mixed on that. On one hand can I understand the reasoning behind this, but on the other can I see situations where someone has an idea, that could also be added, but lacks the knowledge to really add it themself, while others may have the ability too.
Or they just want to spark a discussion on a topic that may be important for the future, but not yet necessary for the project itself (therefore not require active implementation of any kind)...
We shouldn't really stop the posability of brainstorming ideas here, if the topic may not yet be something for the public.

I've answered this in https://github.com/mkdocs/mkdocs-team-space/discussions/3#discussioncomment-9053168 and https://github.com/mkdocs/mkdocs-team-space/discussions/3#discussioncomment-9053192. I would absolutely recommend that we stick to this for now to actually move forward. We have more than 120 open issues – we need to focus on getting this project under control before adding bells and whistles. As you stated before, you'll likely not be able to help out on development, so it is on the rest of us to do so. Thus, please use the public discussion board for now, as you're actively helping us to concentrate on the most pressing things.

**Reply by squidfunk** [reply]
On a side note: I've already invested 2-3h in writing up comments and answers up for this space, which is fine for now. However, I would love to dedicate more of my very limited time to actual work, then discussing what if's 😉

**Reply by Andre601** [reply]
Then I'll honestly take my leave, because my existance here pretty much boils down to proposing "what ifs".

I feel like you and Tim have some differing views here on what and for who this discussions should be it seems.

You seem to focus solely on plugin, theme and extension devs, but have you considered the majority: The enduser using MkDocs?
I am one of these and have ideas that could benefit MkDocs and its userbase, but your focus on plugin, theme and extension dev feedback here makes me feel like this isn't my place and that my inputs are less valued in the end, so I'll leave and focus on public discussions within the mkdocs repository.

Take care.

**Reply by pawamoy** [reply]
If you're still here @Andre601, Martin and I probably have slightly different ways of working, but I believe we share the same vision for the end-goal, and that together we reach a certain balance :slightly_smiling_face: 

Martin does consider the majority (end-users): it is *for the benefit of end-users* that this space was created. We believe it's important to have a space isolated from end-users, where actual plugin authors and MkDocs contributors can work without getting many, many different opinions and suggestions from users, which can drown the most urgent necessities.

That doesn't mean we don't listen to users: we all have access to the public discussions too :slightly_smiling_face: 
It just wouldn't make sense to duplicate or move what happens in the public discussions here: why then would we have created this private space in the first place? It wouldn't have been needed.

Your inputs here are definitely valued, we simply try to stay focused on what we consider the top priorities for MkDocs and the community.

**Reply by squidfunk** [reply]
This space is meant to have different views an discuss them! My comment on using the public discussion board was related to discussing "what if's". The upside is that you'll get a lot more reactions from actual users than in this confined space.

As said, we're currently in a very, very early stage in finding how we want to work and what you read is _my_ take on how I imagine the best way forward is. If the others think that it's a good idea to discuss "what if's" here, that's great and I'm fine with that. Please don't get me wrong: my interest is in pushing this project forward and I'm very interested in using the momentum and positive energy that we generated to actually change something. I might, however, be overly focused on getting sh*t  done right now 😅

> You seem to focus solely on plugin, theme and extension devs, but have you considered the majority: The enduser using MkDocs?

Correct! That's the entire point of this private space. [It is intended for plugin, theme and extension developers](https://github.com/mkdocs/mkdocs/discussions/3622):

> In line with this goal, [@]tomchristie has created a private space (a repository in the MkDocs GitHub organization with the Discussions feature activated) where we can foster better communication and collaboration among community members, by inviting authors of MkDocs themes and plugins, and authors of Markdown extensions. This space will serve as a platform for suggesting ideas for MkDocs and discussing how we can improve MkDocs to better serve plugin authors like ourselves. It will also be a place to discuss about project organization (like issue management, policies, onboarding new maintainers, etc.), while ensuring progress and maintaining focus on our objectives.

I thought @pawamoy invited you because you fall into that category. However, the touchpoint for the enduser and to throw out and discuss ideas is the public discussion board.

**Reply by pawamoy** [reply]
I did initially mention @Andre601 as a plugin author because they [created one](https://github.com/Andre601/mkdocs-pagenav-generator) :slightly_smiling_face: 
Anyway, I don't have a lot more to say, and I agree with @squidfunk's latest comments :blush:

**Reply by lovelydinosaur** [reply]
@squidfunk I don't think there's a massive problem here, but I do need to raise that I have some issues with your style of communication towards others in this community at times. There's several cases where I'd suggest you're coming across as demanding rather than respectful, both towards members in the MkDocs community and also towards our friends over at ReadTheDocs.

I get the impression this is more from a place of a place of wanting to actually push things forwards, rather than being ill-intentioned, but a little more care around how you're using your voice here would be appreciated.

Raising an issue of "toxic behavior" immediately after yourself and Oleh having a critical dispute where I think there was some slightly unprofessional behavior on both sides just seems a bit off-key.

*Okay, awkward conversation done, thanks for listening. The value of your work speaks for itself and you're obviously a hugely welcome member of the MkDocs community, that's not going to change. (☺️)*

**Reply by squidfunk** [reply]
Thanks @tomchristie for being so honest, I'll be very cautious about my choice of words from now on. I sprinkled "IMHO" everywhere and outlined multiple times that my statements are opinions, not decisions. My very goal was to use the energy and momentum to get something done, and I now realize I may have come across too pushy on things. 

I will from now on back-off and let the community of plugin and extension authors decide on things, so that it doesn't appear that I'm pushing some agenda, which I'm not. My only interest is to put MkDocs on a more solid foundation, which I also outlined in the video call we had, and I currently can primarily help out with some time I steal from my other projects. When we talked, I felt that we where very much aligned of what we envisioned MkDocs could become, and how we could somehow upstream and consolidate some of our efforts and plugin and extension maintainers. However, I'm not the one to decide.

Just for a bit of context, because I think it might be helpful to understand me: I'm an entrepreneur at heart (I've never been full employed), and I have seen many projects fall apart because too many ideas get thrown out in the beginning leading to too many possible directions, which more than often ended in analysis paralysis and nothing being done at all. However, this style does not resonate with everyone, which I accept and will from now on cease to follow ✌️

**Reply by pawamoy** [reply]
I feel like we're all a bit sensitive right now. I understand that Martin's messages about toxic behavior can easily be linked to the recent situation with Oleh, however I *know* that they are the result of different situations happening over many years, on different platforms and between different people, not just the recent situation with Oleh.

About the communication style: I agree that Martin can be demanding at times, but I'd say rightfully so. He does want the best for MkDocs, has experience managing Material for MkDocs, and seems to be successful at it. Maybe that's not the right approach here however, as different contributors will always have a different work pace.

Aaaaah, humans are hard :sweat_smile:
